dill/session.py

```diff
@@ -0,0 +1,220 @@
+#!/usr/bin/env python
+#
+# Author: Mike McKerns (mmckerns @caltech and @uqfoundation)
+# Author: Leonardo Gama (@leogama)
+# Copyright (c) 2008-2015 California Institute of Technology.
+# Copyright (c) 2016-2022 The Uncertainty Quantification Foundation.
+# License: 3-clause BSD.  The full license text is available at:
+#  - https://github.com/uqfoundation/dill/blob/master/LICENSE
+"""
+Pickle and restore the intepreter session.
+"""
+
+__all__ = ['dump_session', 'load_session', 'ipython_filter', 'ExcludeRules', 'EXCLUDE', 'INCLUDE']
+
+import re
+import sys
+from copy import copy
+
+from dill import _dill, Pickler, Unpickler
+from ._dill import ModuleType, _import_module, _is_builtin_module, _main_module
+from ._utils import AttrDict, ExcludeRules, Filter, RuleType
+from .settings import settings
+
+# Classes and abstract classes for type hints.
+from io import BytesIO
+from os import PathLike
+from typing import Iterable, NoReturn, Union
+
+EXCLUDE, INCLUDE = RuleType.EXCLUDE, RuleType.INCLUDE
+
+SESSION_IMPORTED_AS_TYPES = tuple([Exception] + [getattr(_dill, name) for name in
+        ('ModuleType', 'TypeType', 'FunctionType', 'MethodType', 'BuiltinMethodType')])
+
+def _module_map():
+    """get map of imported modules"""
+    from collections import defaultdict
+    modmap = AttrDict(by_name=defaultdict(list), by_id=defaultdict(list), top_level={})
+    for modname, module in sys.modules.items():
+        if not isinstance(module, ModuleType):
+            continue
+        if '.' not in modname:
+            modmap.top_level[id(module)] = modname
+        for objname, modobj in module.__dict__.items():
+            modmap.by_name[objname].append((modobj, modname))
+            modmap.by_id[id(modobj)].append((modobj, objname, modname))
+    return modmap
+
+def _lookup_module(modmap, name, obj, main_module):
+    """lookup name or id of obj if module is imported"""
+    for modobj, modname in modmap.by_name[name]:
+        if modobj is obj and sys.modules[modname] is not main_module:
+            return modname, name
+    if isinstance(obj, SESSION_IMPORTED_AS_TYPES):
+        for modobj, objname, modname in modmap.by_id[id(obj)]:
+            if sys.modules[modname] is not main_module:
+                return modname, objname
+    return None, None
+
+def _stash_modules(main_module):
+    modmap = _module_map()
+    newmod = ModuleType(main_module.__name__)
+
+    imported = []
+    imported_as = []
+    imported_top_level = []  # keep separeted for backwards compatibility
+    original = {}
+    for name, obj in vars(main_module).items():
```

Comment:
`vars` < eyeroll... >

Reply:
This was written before that discussion we had...